### PR TITLE
#130. Replicate mark as favourite problem.

### PR DESCRIPTION
**Pull request by MatyashDen**

### CODE REVIEW

File: ios/Codeforces Watcher/Application/features/AppDelegate.swift [hunk 1]
```diff
@@ -55,15 +55,19 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
...
+            $0.titleTextAttributes = [NSAttributedString.Key.foregroundColor: Palette.white,
                                       NSAttributedString.Key.font: Font.textPageTitle]
         }
 
         UITabBar.appearance().run {
             $0.isTranslucent = false
         }
+        
+        UITabBar.appearance().run {
+            $0.itemPositioning = .centered
```
Comment (yevhenii-kanivets) [inline ios/Codeforces Watcher/Application/features/AppDelegate.swift:69]: Move it to declaration on top.

File: ios/Codeforces Watcher/Application/core/design/Palette.swift [hunk 2]
```diff
@@ -9,7 +9,7 @@
 import Foundation
 import UIKit
 
-public class Pallete {
+public class Palette {
```
Comment (yevhenii-kanivets) [inline ios/Codeforces Watcher/Application/core/design/Palette.swift:12]: Wow 👀

File: ios/Codeforces Watcher/Application/features/problems/ProblemsTableViewAdapter.swift [hunk 3]
```diff
@@ -13,43 +13,63 @@ import common
 class ProblemsTableViewAdapter: NSObject, UITableViewDelegate, UITableViewDataSource {
     
     var problems: [Problem] = []
+    var favouriteProblems: [Problem] = []
```
Comment (yevhenii-kanivets) [inline ios/Codeforces Watcher/Application/features/problems/ProblemsTableViewAdapter.swift:16]: I see multiple code-smells here:
- `TableViewAdapter` knows about `store`
- `problems` and `favouriteProblems` server exactly the same purpose, but treated the same way
- `isFavourite` is used in at least 5 places

Generally speaking, you've put the logic, which should exist outside `TableViewAdapter`, inside it. Check the Android implementation and how elegantly it's organized in `ProblemsFragment.newState()`.

File: ios/Codeforces Watcher/Application/features/problems/ProblemsViewController.swift [hunk 4]
```diff
@@ -81,9 +97,27 @@ class ProblemsViewController: UIViewController, StoreSubscriber, UISearchResults
...
     }
+    
+    private func setInteractions() {
+        fabButton.addTarget(self, action: #selector(fabButtonTapped), for: .touchUpInside)
+    }
+    
+    @objc func fabButtonTapped() {
+        var isFavourite = store.state.problems.isFavourite
+        
+        isFavourite = !isFavourite
```
Comment (yevhenii-kanivets) [inline ios/Codeforces Watcher/Application/features/problems/ProblemsViewController.swift:111]: This like looks redundant, you can do the same in `store.dispatch` line.

File: ios/Codeforces Watcher/Application/features/problems/ProblemsViewController.swift [hunk 5]
```diff
@@ -27,7 +32,8 @@ class ProblemsViewController: UIViewController, StoreSubscriber, UISearchResults
 
     override func viewWillAppear(_ animated: Bool) {
         super.viewWillAppear(animated)
-
+        
+        fabButton.isHidden = false
```
Comment (yevhenii-kanivets) [inline ios/Codeforces Watcher/Application/features/problems/ProblemsViewController.swift:36]: `fabButton` is better to be added and controlled in `MainViewController`? Having separate buttons in each sub-controller complicates the code, no?

**Comment by MatyashDen** [comment]
@yevhenii-kanivets 
Could you test press on shareFabView in actions?

File: ios/Codeforces Watcher/Application/features/problems/ProblemsTableViewAdapter.swift [hunk 6]
```diff
@@ -13,43 +13,48 @@ import common
...
     var problems: [Problem] = []
+    var visibleProblems: [Problem] = []
 
     var onProblemClick: ((String, String) -> ())?
 
     func numberOfSections(in tableView: UITableView) -> Int {
         return 1
     }
+    
+    var isFavourite = false
```
Comment (yevhenii-kanivets) [inline ios/Codeforces Watcher/Application/features/problems/ProblemsTableViewAdapter.swift:24]: It's better to set string of "empty problems" explanation, than boolean here. (So adapter is as dummy as possible).

File: ios/Codeforces Watcher/Application/features/problems/ProblemsTableViewAdapter.swift [hunk 7]
```diff
@@ -13,43 +13,48 @@ import common
 class ProblemsTableViewAdapter: NSObject, UITableViewDelegate, UITableViewDataSource {
     
     var problems: [Problem] = []
+    var visibleProblems: [Problem] = []
```
Comment (yevhenii-kanivets) [inline ios/Codeforces Watcher/Application/features/problems/ProblemsTableViewAdapter.swift:16]: Again, this second array is too much responsibility for a dummy adapter 😉If you really need to store all `problems` locally, store it in VC.

File: ios/Codeforces Watcher/Application/features/problems/ProblemsViewController.swift [hunk 8]
```diff
@@ -81,9 +84,21 @@ class ProblemsViewController: UIViewController, StoreSubscriber, UISearchResults
...
+    
+    override func fabButtonTapped() {
+        let isFavourite = !store.state.problems.isFavourite
+        
+        store.dispatch(action: ProblemsActions.ChangeTypeProblems(isFavourite: isFavourite))
+        
+        updateFabButton(isFavourite)
+    }
+
+    private func updateFabButton(_ isFavourite: Bool) {
```
Comment (yevhenii-kanivets) [inline ios/Codeforces Watcher/Application/features/problems/ProblemsViewController.swift:99]: Isn't it better to do in `doNewState` when `isFavourite` is changing there?

**Comment by yevhenii-kanivets** [comment]
@MatyashDen share FAB in Actions tab works, but the text is not localized (I see `cw_share_message`). Also on my iPhone 5S FAB is overlapping tabs in tab bottom bar. On Android we've hacked it by adding empty 5th tab (actually 3rd) tab, so all other tabs are moved to edges. You can try smth similar :)

**Pull request merged**